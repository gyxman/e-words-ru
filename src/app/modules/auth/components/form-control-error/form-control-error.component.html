<ng-container *ngIf="control.touched && control.errors">
    <mat-error
        *ngIf="control.errors.required"
        automation-id="error-required"
        class="error"
    >
        Поле обязательно для заполнения
    </mat-error>
    <mat-error *ngIf="control.errors.email" automation-id="error-email" class="error">
        Неверный формат E-mail
    </mat-error>
    <mat-error
        *ngIf="control.errors.minlength"
        automation-id="error-minlength"
        class="error"
    >
        Пароль должен быть не менее
        {{ control.errors.minlength.requiredLength }} символов. Сейчас
        {{ control.errors.minlength.actualLength }} символов
    </mat-error>
</ng-container>
