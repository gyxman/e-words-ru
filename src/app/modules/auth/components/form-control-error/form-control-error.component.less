@import '../../styles/index';

:host {
    display: block;
}

.error {
    .font(13px, 20px);
}
