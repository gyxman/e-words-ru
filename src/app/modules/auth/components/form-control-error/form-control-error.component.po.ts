import {DebugElement} from '@angular/core';
import {PageObject} from '../../../../testing/page-object';

export class FormControlErrorComponentPo<T> extends PageObject<T> {
    get errorRequired(): DebugElement {
        return this.getByAutomationId('error-required');
    }

    get errorEmail(): DebugElement {
        return this.getByAutomationId('error-email');
    }

    get errorMinlength(): DebugElement {
        return this.getByAutomationId('error-minlength');
    }
}
