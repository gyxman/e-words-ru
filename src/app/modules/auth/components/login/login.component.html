<form automation-id="form" class="form" [formGroup]="form" (ngSubmit)="onSubmit()">
    <h2 class="form-title">
        Войти в тренажер
    </h2>
    <div class="form-content">
        <div class="form-row">
            <mat-form-field class="form-field">
                <mat-label>E-mail</mat-label>
                <input
                    automation-id="email-input"
                    matInput
                    placeholder="Введите e-mail"
                    [formControl]="emailControl"
                />
                <app-form-control-error
                    class="form-error"
                    [control]="emailControl"
                ></app-form-control-error>
            </mat-form-field>
        </div>
        <div class="form-row">
            <mat-form-field class="form-field">
                <mat-label>Пароль</mat-label>
                <input
                    automation-id="password-input"
                    type="password"
                    matInput
                    placeholder="Введите пароль"
                    [formControl]="passwordControl"
                />
                <app-form-control-error
                    class="form-error"
                    [control]="passwordControl"
                ></app-form-control-error>
            </mat-form-field>
        </div>
        <div class="form-row form-row_actions">
            <button
                automation-id="submit-button"
                mat-raised-button
                color="primary"
                class="button"
                type="submit"
                [disabled]="form.invalid"
            >
                Войти
            </button>
        </div>
    </div>
</form>
