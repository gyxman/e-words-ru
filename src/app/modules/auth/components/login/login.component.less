@import '../../styles/index';

:host {
    display: block;
}
