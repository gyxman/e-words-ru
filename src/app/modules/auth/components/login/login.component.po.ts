import {PageObject} from '../../../../testing/page-object';
import {DebugElement} from '@angular/core';

export class LoginComponentPo<T> extends PageObject<T> {
    get form(): DebugElement {
        return this.getByAutomationId('form');
    }

    get emailInput(): HTMLInputElement {
        return this.getByAutomationId('email-input').nativeElement;
    }

    get passwordInput(): HTMLInputElement {
        return this.getByAutomationId('password-input').nativeElement;
    }

    get submitButton(): DebugElement {
        return this.getByAutomationId('submit-button');
    }
}
