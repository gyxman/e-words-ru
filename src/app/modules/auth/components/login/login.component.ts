import {ChangeDetectionStrategy, Component} from '@angular/core';
import {FormControl, FormGroup, Validators} from '@angular/forms';
import {AuthFacadeService} from '../../services/auth-facade.service';

@Component({
    selector: 'app-login',
    templateUrl: './login.component.html',
    styleUrls: ['./login.component.less'],
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class LoginComponent {
    form = new FormGroup({
        email: new FormControl(null, [Validators.email, Validators.required]),
        password: new FormControl(null, [Validators.minLength(8), Validators.required]),
    });

    readonly showLoader$ = this.authFacadeService.showLoader$;

    constructor(private readonly authFacadeService: AuthFacadeService) {}

    get emailControl(): FormControl {
        return this.form.get('email') as FormControl;
    }

    get passwordControl(): FormControl {
        return this.form.get('password') as FormControl;
    }

    onSubmit() {
        const data = this.form.value;

        this.authFacadeService.signInWithEmail(data);
    }
}
