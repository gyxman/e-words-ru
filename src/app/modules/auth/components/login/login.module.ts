import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {LoginComponent} from './login.component';
import {ReactiveFormsModule} from '@angular/forms';
import {MatInputModule} from '@angular/material/input';
import {FormControlErrorModule} from '../form-control-error/form-control-error.module';
import {MatButtonModule} from '@angular/material/button';

@NgModule({
    imports: [
        CommonModule,
        ReactiveFormsModule,
        MatInputModule,
        MatButtonModule,
        FormControlErrorModule,
    ],
    declarations: [LoginComponent],
    exports: [LoginComponent],
})
export class LoginModule {}
