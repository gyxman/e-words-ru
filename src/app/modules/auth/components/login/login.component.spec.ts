import {ComponentFixture, TestBed} from '@angular/core/testing';
import {LoginComponent} from './login.component';
import {MockModule} from 'ng-mocks';
import {ReactiveFormsModule} from '@angular/forms';
import {MatInputModule} from '@angular/material/input';
import {MatButtonModule} from '@angular/material/button';
import {FormControlErrorModule} from '../form-control-error/form-control-error.module';
import {AuthFacadeService} from '../../services/auth-facade.service';
import {deepEqual, instance, mock, verify} from 'ts-mockito';
import {LoginComponentPo} from './login.component.po';

describe('LoginComponent - компонент входа в приложение', () => {
    let component: LoginComponent;
    let fixture: ComponentFixture<LoginComponent>;
    let pageObject: LoginComponentPo<LoginComponent>;
    let authFacadeServiceMock: AuthFacadeService;

    beforeEach(() => {
        authFacadeServiceMock = mock(AuthFacadeService);
    });

    beforeEach(() => {
        TestBed.configureTestingModule({
            declarations: [LoginComponent],
            imports: [
                ReactiveFormsModule,
                MockModule(MatInputModule),
                MockModule(MatButtonModule),
                MockModule(FormControlErrorModule),
            ],
            providers: [
                {
                    provide: AuthFacadeService,
                    useFactory: () => instance(authFacadeServiceMock),
                },
            ],
        }).compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(LoginComponent);
        component = fixture.componentInstance;
        pageObject = new LoginComponentPo(fixture);
    });

    it('Если компонент загрузился, то все элементы отрисовались', () => {
        // act
        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('Если пользователь открывает страниицу логина, то кнопка "Войти" недоступна', () => {
        // act
        fixture.detectChanges();

        // assert
        expect(pageObject.submitButton.attributes['ng-reflect-disabled']).toBe('true');
    });

    it('Если пользователь открывает страниицу логина, заполняет e-mail валидным значением, кнопка "Войти" недоступна', () => {
        // arrange
        fixture.detectChanges();

        // act
        pageObject.emailInput.value = 'test@test.ru';
        pageObject.emailInput.dispatchEvent(new Event('input'));

        fixture.detectChanges();

        // assert
        expect(pageObject.submitButton.attributes['ng-reflect-disabled']).toBe('true');
    });

    it('Если пользователь открывает страниицу логина, заполняет пароль валидным значением, кнопка "Войти" недоступна', () => {
        // arrange
        fixture.detectChanges();

        // act
        pageObject.passwordInput.value = 'testPassword';
        pageObject.passwordInput.dispatchEvent(new Event('input'));

        fixture.detectChanges();

        // assert
        expect(pageObject.submitButton.attributes['ng-reflect-disabled']).toBe('true');
    });

    it(`Если пользователь открывает страниицу логина, заполняет e-mail невалидным значением,
        а пароль валидным значением, кнопка "Войти" недоступна`, () => {
        // arrange
        fixture.detectChanges();

        // act
        pageObject.emailInput.value = 'test';
        pageObject.emailInput.dispatchEvent(new Event('input'));

        pageObject.passwordInput.value = 'testPassword';
        pageObject.passwordInput.dispatchEvent(new Event('input'));

        fixture.detectChanges();

        // assert
        expect(pageObject.submitButton.attributes['ng-reflect-disabled']).toBe('true');
    });

    it(`Если пользователь открывает страниицу логина, заполняет e-mail валидным значением,
        а пароль невалидным значением, кнопка "Войти" недоступна`, () => {
        // arrange
        fixture.detectChanges();

        // act
        pageObject.emailInput.value = 'test@test.ru';
        pageObject.emailInput.dispatchEvent(new Event('input'));

        pageObject.passwordInput.value = 'test';
        pageObject.passwordInput.dispatchEvent(new Event('input'));

        fixture.detectChanges();

        // assert
        expect(pageObject.submitButton.attributes['ng-reflect-disabled']).toBe('true');
    });

    it('Если пользователь открывает страниицу логина, заполняет e-mail и пароль валидными значениями, кнопка "Войти" доступна', () => {
        // arrange
        fixture.detectChanges();

        // act
        pageObject.emailInput.value = 'test@test.ru';
        pageObject.emailInput.dispatchEvent(new Event('input'));

        pageObject.passwordInput.value = 'testPassword';
        pageObject.passwordInput.dispatchEvent(new Event('input'));

        fixture.detectChanges();

        // assert
        expect(pageObject.submitButton.attributes['ng-reflect-disabled']).toBe('false');
    });

    it('Если пользователь отправляет форму, то отправляем запрос на авторизацию данными, которые ввел пользователь', () => {
        // act
        fixture.detectChanges();

        pageObject.emailInput.value = 'test@test.ru';
        pageObject.emailInput.dispatchEvent(new Event('input'));

        pageObject.passwordInput.value = 'testPassword';
        pageObject.passwordInput.dispatchEvent(new Event('input'));

        pageObject.form.triggerEventHandler('ngSubmit', null);

        // assert
        verify(
            authFacadeServiceMock.signInWithEmail(
                deepEqual({
                    email: 'test@test.ru',
                    password: 'testPassword',
                }),
            ),
        ).once();
    });
});
