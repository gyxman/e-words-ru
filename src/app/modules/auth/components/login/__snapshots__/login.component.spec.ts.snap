// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent - компонент входа в приложение Если компонент загрузился, то все элементы отрисовались 1`] = `
<div
  id="root0"
>
  <form
    automation-id="form"
    class="form ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <h2
      class="form-title"
    >
       Войти в тренажер 
    </h2>
    <div
      class="form-content"
    >
      <div
        class="form-row"
      >
        <mat-form-field
          class="form-field"
        >
          <mat-label>
            E-mail
          </mat-label>
          <input
            automation-id="email-input"
            class="ng-untouched ng-pristine ng-invalid"
            matinput=""
            placeholder="Введите e-mail"
          />
          <app-form-control-error
            class="form-error"
          />
        </mat-form-field>
      </div>
      <div
        class="form-row"
      >
        <mat-form-field
          class="form-field"
        >
          <mat-label>
            Пароль
          </mat-label>
          <input
            automation-id="password-input"
            class="ng-untouched ng-pristine ng-invalid"
            matinput=""
            placeholder="Введите пароль"
            type="password"
          />
          <app-form-control-error
            class="form-error"
          />
        </mat-form-field>
      </div>
      <div
        class="form-row form-row_actions"
      >
        <button
          automation-id="submit-button"
          class="button"
          color="primary"
          mat-raised-button=""
          type="submit"
        >
           Войти 
        </button>
      </div>
    </div>
  </form>
</div>
`;
